refinenet/refinenet.py

```diff
@@ -0,0 +1,253 @@
+            low_res_adap_conv_out = None
+            if low_res_input is not None:
+                low_res_adap_conv_out = residual_conv_unit(low_res_input)
+                low_res_adap_conv_out = residual_conv_unit(low_res_adap_conv_out)
+            high_res_adap_conv_out = residual_conv_unit(high_res_input)
+            high_res_adap_conv_out = residual_conv_unit(high_res_adap_conv_out)
+
+            # Multi-resolution Fusion
+            mrf_out = multi_resolution_fusion(low_res_adap_conv_out, high_res_adap_conv_out)
+
+            # Chained Residual Pooling
+            crp_out = chained_residual_pooling(mrf_out)
+
+            out_conv = residual_conv_unit(crp_out)
+
+            return out_conv
+
+
+        # build ResNet
+        nets, end_points = resnet_50(inputs=processed_image,
+                                     num_classes=self.num_classes,
+                                     is_training=self.is_training,
+                                     scope="resnet_v2")
+
+        # get outputs from ResNet that we want to refine on
+        multi_res_outs = [end_points['resnet_v2/block1/unit_2/bottleneck_v2'],
+                          end_points['resnet_v2/block2/unit_3/bottleneck_v2'],
+                          end_points['resnet_v2/block3/unit_5/bottleneck_v2'],
+                          end_points['resnet_v2/block4']]
+
+        # refine outputs recursively
+        refine_out = None
+        for i in [3, 2, 1, 0]:
+            refine_out = refine_block(refine_out, multi_res_outs[i])
+
+
+        # two more rcus right before the final scoring layer
+        for i in range(2):        
+            refine_out = residual_conv_unit(refine_out)
+        
+        # scoring layer
+        logits = tf.layers.conv2d(refine_out, filters=self.num_classes,
+                                      kernel_size=3, strides=1, padding="same",
+                                      data_format=self.data_format)
+
+        if self.data_format == "channels_first":
+            logits = tf.transpose(logits, [0, 2, 3, 1])
+
+        # bilinearly up-sample to match the dimension of the input image
+        logits = tf.image.resize_images(logits, self.input_shape)
+
+        annotation_pred = tf.argmax(logits, axis=3, name="prediction", output_type=tf.int32)
+
+        return tf.expand_dims(annotation_pred, dim=3), logits
+
+    def train(self, loss_val, var_list):
+        """Add training operations to computational graph."""
+        self.global_step = tf.get_variable(name="global_step", dtype=tf.int32, initializer=0)
+        update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
+        with tf.control_dependencies(update_ops):
+            optimizer = tf.train.AdamOptimizer(self.learning_rate)
+            grads = optimizer.compute_gradients(loss_val, var_list=var_list)
+            return optimizer.apply_gradients(grads, self.global_step)
+
+if __name__ == "__main__":
+    """ run this main function to test the model """
```

Comment:
The default main procedure should be training/testing on a real dataset; the contents of `__main__` right now would be better off as a separate test function. Is this here because of difficulty in getting the paper's datasets?

Reply:
The main function is solely to make sure the model can build and run.

PASCAL VOC is not hard to get, but ImageNet is a little bit. Need more time to integrate data preprocessing and training/testing functionalities to this module.

Reply:
Sounds good, this is good for now, but we do need to add a public dataset so others can verify

Reply:
Yes, I'll work on this.